--- showconsole-1.10/blogd.c~	2010-08-30 20:34:32.000000000 +0300
+++ showconsole-1.10/blogd.c	2010-08-30 20:35:03.281749868 +0300
@@ -220,6 +220,7 @@
 {
     int fd, fd2, flags;
     int ptm, pts, cntrtty = 1;
+    int quiet = 0;
     const pid_t ppid = getppid();
     const pid_t pgrp = getpgid(ppid);
     char ptsname[NAME_MAX+1];
@@ -193,7 +194,14 @@
     myname = basename(*argv);
 
     if (argc > 2)
-	error("Too many args; usage: %s [/dev/tty<X>]\n", myname);
+	error("Too many args; usage: %s [-q] [/dev/tty<X>]\n", myname);
+
+    if (argc > 1 && strcmp(argv[1], "-q") == 0) {
+	quiet = 1;
+	--argc;
+	if (argc > 2)
+	    argv[1] = argv[2];
+    }
 
     if (argc == 2)
 	tty = argv[1];
@@ -362,6 +362,7 @@
 	close(fd);
 	if (fd2 > 0)
 	    close(fd2);
+	if (!quiet)
 	fprintf(stdout, "\rBoot logging started on %s(%s) at %.24s\n", tty, name, stt);
 	fflush(stdout);
 	exit(0);
